### PR TITLE
TMA-775: deterministic lookup of label values

### PR DESCRIPTION
**Pull request by kubamahnert**

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/665/ : SUCCESS in 45s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/466/ : FAILURE in 32s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/774/ : SUCCESS in 33s

File: lib/gooddata/models/metadata/label.rb [hunk 1]
```diff
@@ -42,29 +42,16 @@ def find_element_value(element_id)
...
-        end
+    def get_valid_elements(*args)
+      results, params = valid_elements(*args)
+      # TMA-775 - the validElements API can possibly return more matches than requested (usually 1)
+      # so we do a preliminary first request to check and then increase the limit if needed
+      if results['validElements']['paging']['total'].to_i != params[:limit]
+        params[:limit] = 100_000
+        results, _ = valid_elements params
+        results['validElements']['items'] = results['validElements']['items'].reject do |i|
+          i['element']['title'] != params[:filter]
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Calls 'params[:filter]' 2 times (DuplicateMethodCall)

File: lib/gooddata/models/metadata/label.rb [hunk 2]
```diff
@@ -42,29 +42,16 @@ def find_element_value(element_id)
...
-      if status_url
-        results = client.poll_on_response(status_url) do |body|
-          status = body['taskState'] && body['taskState']['status']
-          status == 'RUNNING' || status == 'PREPARED'
-        end
+    def get_valid_elements(*args)
+      results, params = valid_elements(*args)
+      # TMA-775 - the validElements API can possibly return more matches than requested (usually 1)
+      # so we do a preliminary first request to check and then increase the limit if needed
+      if results['validElements']['paging']['total'].to_i != params[:limit]
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Calls 'results['validelements']' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/models/metadata/label.rb [hunk 3]
```diff
@@ -122,6 +109,36 @@ def attribute
...
+      end
+
+      results = client.post(final_url, 'validElementsRequest' => request_payload)
+
+      # Implementation of polling is based on
+      # https://opengrok.intgdc.com/source/xref/gdc-backend/src/test/java/com/gooddata/service/dao/ValidElementsDaoTest.java
+      status_url = results['uri']
+      if status_url
+        results = client.poll_on_response(status_url) do |body|
+          status = body['taskState'] && body['taskState']['status']
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Calls 'body['taskstate']' 2 times (DuplicateMethodCall)

File: lib/gooddata/models/metadata/label.rb [hunk 4]
```diff
@@ -122,6 +109,36 @@ def attribute
     def attribute_uri
       content['formOf']
     end
+
+    private
+    def valid_elements(url_or_params = {}, request_payload = {})
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Has approx 13 statements (TooManyStatements)

File: lib/gooddata/models/metadata/label.rb [hunk 5]
```diff
@@ -42,29 +42,16 @@ def find_element_value(element_id)
...
-          status == 'RUNNING' || status == 'PREPARED'
-        end
+    def get_valid_elements(*args)
+      results, params = valid_elements(*args)
+      # TMA-775 - the validElements API can possibly return more matches than requested (usually 1)
+      # so we do a preliminary first request to check and then increase the limit if needed
+      if results['validElements']['paging']['total'].to_i != params[:limit]
+        params[:limit] = 100_000
+        results, _ = valid_elements params
+        results['validElements']['items'] = results['validElements']['items'].reject do |i|
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Has the variable name 'i' (UncommunicativeVariableName)

Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Style/MultilineIfModifier: Favor a normal if-statement over a modifier clause in a multiline statement.

File: lib/gooddata/models/metadata/label.rb [hunk 6]
```diff
@@ -122,6 +109,36 @@ def attribute
...
+      final_url = url_or_params
+
+      if url_or_params.is_a?(Hash)
+        default_params = {
+          limit: 1,
+          offset: 0,
+          order: 'asc'
+        }
+        params = default_params.merge(url_or_params)
+        query_params = params.map { |x, v| "#{x}=#{CGI.escape(v.to_s)}" }.reduce { |acc, elem| "#{acc}&#{elem}" }
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Has the variable name 'v' (UncommunicativeVariableName)

Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Has the variable name 'x' (UncommunicativeVariableName)

File: spec/helpers/appstore_project_helper.rb [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+          report.save
+        end
+      end
+
+      def create_dashboards
+        GoodData.with_project @project.pid do
+          dashboards = JSON.parse(File.read(DASHBOARDS_FILE)).to_hash['dashboards']
+          dashboards.each do |d|
+            dashboard = GoodData::Dashboard.create(
+              { title: d['title'] },
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Calls 'd['title']' 2 times (DuplicateMethodCall)

File: spec/helpers/appstore_project_helper.rb [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+              client: client, project: @project
+            )
+            d['tabs'].each do |t|
+              tab = dashboard.add_tab(title: t['title'])
+              t['items'].each do |i|
+                tab.add_report_item(report: GoodData::Report[i['identifier']],
+                                    position_x: i['position']['x'],
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Calls 'i['position']' 2 times (DuplicateMethodCall)

File: spec/helpers/appstore_project_helper.rb [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+                                    position_y: i['position']['y'],
+                                    size_x: i['size']['width'],
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Calls 'i['size']' 2 times (DuplicateMethodCall)

File: spec/helpers/appstore_project_helper.rb [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        project.attributes(COMPUTED_ATTRIBUTE_ID).drill_down(project.attributes('attr.csv_policies.customer'))
+      end
+
+      def create_reports
+        GoodData.with_project @project.pid do |p|
+          reports = JSON.parse(File.read(REPORTS_FILE)).to_hash['reports']
+          reports.each do |r|
+            report = p.add_report(
+              title: r['title'],
+              left: r['left'].map { |o| GoodData::MdObject[o] },
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Calls 'gooddata::mdobject[o]' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Contains iterators nested 3 deep (NestedIterators)

Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Has the variable name 'o' (UncommunicativeVariableName)

File: spec/helpers/appstore_project_helper.rb [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+
+        project.attributes(COMPUTED_ATTRIBUTE_ID).drill_down(project.attributes('attr.csv_policies.customer'))
+      end
+
+      def create_reports
+        GoodData.with_project @project.pid do |p|
+          reports = JSON.parse(File.read(REPORTS_FILE)).to_hash['reports']
+          reports.each do |r|
+            report = p.add_report(
+              title: r['title'],
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Calls 'r['title']' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Refers to 'r' more than self (maybe move it to another class?) (FeatureEnvy)

File: spec/helpers/appstore_project_helper.rb [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+              left: r['left'].map { |o| GoodData::MdObject[o] },
+              top: r['top'].map { |o| GoodData::MdObject[o] }
+            )
+            report_name = r['title'].to_s.downcase.gsub(/\s/, '.')
+            report.identifier = "report.#{report_name}"
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Calls 'report.identifier = "report.#{report_name}"' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Refers to 'report' more than self (maybe move it to another class?) (FeatureEnvy)

File: spec/helpers/appstore_project_helper.rb [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+            report.save
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Calls 'report.save' 2 times (DuplicateMethodCall)

File: spec/helpers/appstore_project_helper.rb [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+          dashboards = JSON.parse(File.read(DASHBOARDS_FILE)).to_hash['dashboards']
+          dashboards.each do |d|
+            dashboard = GoodData::Dashboard.create(
+              { title: d['title'] },
+              client: client, project: @project
+            )
+            d['tabs'].each do |t|
+              tab = dashboard.add_tab(title: t['title'])
+              t['items'].each do |i|
+                tab.add_report_item(report: GoodData::Report[i['identifier']],
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Refers to 'i' more than self (maybe move it to another class?) (FeatureEnvy)

File: spec/helpers/appstore_project_helper.rb [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+          )
+        end
+      end
+
+      def create_metrics
+        GoodData.with_project @project.pid do |project|
+          metrics = JSON.parse(File.read(METRICS_FILE)).to_hash
+          metrics.each do |title, maql|
+            metric = project.add_measure maql, title: title
+            metric.identifier = "metric.#{title.to_s.downcase.gsub(/\s/, '.')}"
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Refers to 'metric' more than self (maybe move it to another class?) (FeatureEnvy)

File: spec/helpers/appstore_project_helper.rb [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+          project: @project,
+          client: client,
+          client_id: opts[:client_id],
+          output_stage_prefix: opts[:output_stage_prefix]
+        )
+      end
+
+      def deploy_processes(ads)
+        GoodData.with_project @project.pid do |project|
+          cc_process = project.deploy_process(
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Refers to 'project' more than self (maybe move it to another class?) (FeatureEnvy)

File: spec/helpers/appstore_project_helper.rb [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+        dataset.add_tag('dataset')
+        dataset.save
+
+        fact = @project.facts(FACT_IDENTIFIER)
+        fact.add_tag('fact')
+        fact.save
+      end
+
+      def ensure_user(login, domain)
+        user = domain.users(login)
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Refers to 'domain' more than self (maybe move it to another class?) (FeatureEnvy)

File: spec/helpers/appstore_project_helper.rb [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+        GoodData.with_project @project.pid do
+          dashboards = JSON.parse(File.read(DASHBOARDS_FILE)).to_hash['dashboards']
+          dashboards.each do |d|
+            dashboard = GoodData::Dashboard.create(
+              { title: d['title'] },
+              client: client, project: @project
+            )
+            d['tabs'].each do |t|
+              tab = dashboard.add_tab(title: t['title'])
+              t['items'].each do |i|
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Contains iterators nested 3 deep (NestedIterators)

Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Has the variable name 'i' (UncommunicativeVariableName)

File: spec/helpers/appstore_project_helper.rb [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+            de_synchronize_all: true,
+            state: 'DISABLED'
+          )
+        end
+      end
+
+      def create_metrics
+        GoodData.with_project @project.pid do |project|
+          metrics = JSON.parse(File.read(METRICS_FILE)).to_hash
+          metrics.each do |title, maql|
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Contains iterators nested 2 deep (NestedIterators)

File: spec/helpers/appstore_project_helper.rb [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+            metric = project.add_measure maql, title: title
+            metric.identifier = "metric.#{title.to_s.downcase.gsub(/\s/, '.')}"
+            metric.add_tag('metric') if metric.identifier == PRODUCTION_TAGGED_METRIC
+            metric.save
+          end
+        end
+      end
+
+      def create_computed_attributes
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Has approx 7 statements (TooManyStatements)

File: spec/helpers/appstore_project_helper.rb [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+
+          ca_report_title = 'Report contains CA'
+          report = p.add_report(title: ca_report_title, left: GoodData::MdObject[COMPUTED_ATTRIBUTE_ID])
+          report_name = ca_report_title.downcase.gsub(/\s/, '.')
+          report.identifier = "report.#{report_name}"
+          report.save
+        end
+      end
+
+      def create_dashboards
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Has approx 13 statements (TooManyStatements)

File: spec/helpers/appstore_project_helper.rb [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+            ruby_schedule,
+            '',
+            dataload_datasets: [ DATASET_IDENTIFIER ],
+            de_synchronize_all: true,
+            state: 'DISABLED'
+          )
+        end
+      end
+
+      def create_metrics
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Has approx 7 statements (TooManyStatements)

File: spec/helpers/appstore_project_helper.rb [hunk 23]
```diff
@@ -0,0 +1,214 @@
...
+        unless new_blueprint.valid?
+          pp new_blueprint.validate
+          raise 'Cannot create computed attribute'
+        end
+        project.update_from_blueprint(new_blueprint)
+
+        project.attributes(COMPUTED_ATTRIBUTE_ID).drill_down(project.attributes('attr.csv_policies.customer'))
+      end
+
+      def create_reports
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Has approx 14 statements (TooManyStatements)

File: spec/helpers/appstore_project_helper.rb [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+            end
+            dashboard_name = d['title'].to_s.downcase.gsub(/\s/, '.')
+            dashboard.identifier = "dashboard.#{dashboard_name}"
+            dashboard.add_tag('dashboard')
+            dashboard.save
+          end
+        end
+      end
+
+      def create_tag_for_fact_n_dataset
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Has approx 6 statements (TooManyStatements)

File: spec/helpers/appstore_project_helper.rb [hunk 25]
```diff
@@ -0,0 +1,214 @@
...
+        @project.create_output_stage(
+          ads,
+          project: @project,
+          client: client,
+          client_id: opts[:client_id],
+          output_stage_prefix: opts[:output_stage_prefix]
+        )
+      end
+
+      def deploy_processes(ads)
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Has approx 8 statements (TooManyStatements)

File: spec/helpers/appstore_project_helper.rb [hunk 26]
```diff
@@ -0,0 +1,214 @@
...
+    class AppstoreProjectHelper
+      attr_reader :project, :client
+
+      class << self
+        def create(opts = { client: GoodData.connection })
+          new_project = opts[:client].create_project(opts)
+          AppstoreProjectHelper.new(new_project, opts)
+        end
+
+        def suffix
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Has approx 7 statements (TooManyStatements)

File: spec/helpers/appstore_project_helper.rb [hunk 27]
```diff
@@ -0,0 +1,214 @@
...
+          report_name = ca_report_title.downcase.gsub(/\s/, '.')
+          report.identifier = "report.#{report_name}"
+          report.save
+        end
+      end
+
+      def create_dashboards
+        GoodData.with_project @project.pid do
+          dashboards = JSON.parse(File.read(DASHBOARDS_FILE)).to_hash['dashboards']
+          dashboards.each do |d|
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Has the variable name 'd' (UncommunicativeVariableName)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/666/ : SUCCESS in 16s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/467/ : FAILURE in 54s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/775/ : SUCCESS in 36s

File: lib/gooddata/models/metadata/label.rb [hunk 28]
```diff
@@ -42,28 +42,17 @@ def find_element_value(element_id)
...
-          status = body['taskState'] && body['taskState']['status']
-          status == 'RUNNING' || status == 'PREPARED'
+    def get_valid_elements(*args)
+      results, = valid_elements(*args)
+      # TMA-775 - the validElements API can possibly return more matches than requested (usually 1)
+      # so we do a preliminary first request to check and then increase the limit if needed
+      if results['validElements']['paging']['total'].to_i != params[:limit]
+        params[:limit] = 100_000
+        results, _ = valid_elements params
+        if params[:filter]
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Calls 'params[:filter]' 2 times (DuplicateMethodCall)

File: lib/gooddata/models/metadata/label.rb [hunk 29]
```diff
@@ -42,28 +42,17 @@ def find_element_value(element_id)
...
-      status_url = results['uri']
-      if status_url
-        results = client.poll_on_response(status_url) do |body|
-          status = body['taskState'] && body['taskState']['status']
-          status == 'RUNNING' || status == 'PREPARED'
+    def get_valid_elements(*args)
+      results, = valid_elements(*args)
+      # TMA-775 - the validElements API can possibly return more matches than requested (usually 1)
+      # so we do a preliminary first request to check and then increase the limit if needed
+      if results['validElements']['paging']['total'].to_i != params[:limit]
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Calls 'results['validelements']' 3 times (DuplicateMethodCall)

File: lib/gooddata/models/metadata/label.rb [hunk 30]
```diff
@@ -42,28 +42,17 @@ def find_element_value(element_id)
...
-      results = client.post(final_url, 'validElementsRequest' => request_payload)
-
-      # Implementation of polling is based on
-      # https://opengrok.intgdc.com/source/xref/gdc-backend/src/test/java/com/gooddata/service/dao/ValidElementsDaoTest.java
-      status_url = results['uri']
-      if status_url
-        results = client.poll_on_response(status_url) do |body|
-          status = body['taskState'] && body['taskState']['status']
-          status == 'RUNNING' || status == 'PREPARED'
+    def get_valid_elements(*args)
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Has approx 6 statements (TooManyStatements)

File: lib/gooddata/models/metadata/label.rb [hunk 31]
```diff
@@ -122,6 +111,37 @@ def attribute
     def attribute_uri
       content['formOf']
     end
+
+    private
+
+    def valid_elements(url_or_params = {}, request_payload = {})
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Has approx 13 statements (TooManyStatements)

File: spec/helpers/appstore_project_helper.rb [hunk 32]
```diff
@@ -0,0 +1,221 @@
...
+            dashboard = GoodData::Dashboard.create(
+              { title: d['title'] },
+              client: client, project: @project
+            )
+            d['tabs'].each do |t|
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Has the variable name 't' (UncommunicativeVariableName)

File: spec/helpers/appstore_project_helper.rb [hunk 33]
```diff
@@ -0,0 +1,221 @@
...
+          pp new_blueprint.validate
+          raise 'Cannot create computed attribute'
+        end
+        project.update_from_blueprint(new_blueprint)
+
+        project.attributes(COMPUTED_ATTRIBUTE_ID).drill_down(project.attributes('attr.csv_policies.customer'))
+      end
+
+      def create_reports
+        GoodData.with_project @project.pid do |p|
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Has the variable name 'p' (UncommunicativeVariableName)

File: spec/helpers/appstore_project_helper.rb [hunk 34]
```diff
@@ -0,0 +1,221 @@
...
+          reports = JSON.parse(File.read(REPORTS_FILE)).to_hash['reports']
+          reports.each do |r|
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Has the variable name 'r' (UncommunicativeVariableName)

File: spec/helpers/constants.rb [hunk 35]
```diff
@@ -0,0 +1,32 @@
+module GoodData
+  module Helpers
```
Comment (yenkins) [inline spec/helpers/constants.rb]: Has 21 constants (TooManyConstants)

File: lib/gooddata/models/metadata/label.rb [hunk 36]
```diff
@@ -42,28 +42,17 @@ def find_element_value(element_id)
...
-        results = client.poll_on_response(status_url) do |body|
-          status = body['taskState'] && body['taskState']['status']
-          status == 'RUNNING' || status == 'PREPARED'
+    def get_valid_elements(*args)
+      results, = valid_elements(*args)
+      # TMA-775 - the validElements API can possibly return more matches than requested (usually 1)
+      # so we do a preliminary first request to check and then increase the limit if needed
+      if results['validElements']['paging']['total'].to_i != params[:limit]
+        params[:limit] = 100_000
+        results, _ = valid_elements params
```
Comment (yenkins) [inline lib/gooddata/models/metadata/label.rb]: Style/TrailingUnderscoreVariable: Do not use trailing `_`s in parallel assignment. Prefer `results, = valid_elements params`.

File: spec/helpers/appstore_project_helper.rb [hunk 37]
```diff
@@ -0,0 +1,221 @@
+# encoding: UTF-8
+#
+# Copyright (c) 2010-2018 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Style/EmptyLines: Extra blank line detected.

File: spec/helpers/appstore_project_helper.rb [hunk 38]
```diff
@@ -0,0 +1,221 @@
...
+          suffix
+        end
+      end
+
+      def initialize(project, opts = { client: GoodData.connection })
+        @client = opts[:client]
+        @project = project if project.instance_of? GoodData::Project
+        @project = client.projects(project) if project.instance_of? String
+        GoodData.logger.info "Project ID: #{@project.obj_id}"
+        raise 'ProjectHelper::initialize the parameter "project" '\
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Style/MultilineIfModifier: Favor a normal unless-statement over a modifier clause in a multiline statement.

File: spec/helpers/appstore_project_helper.rb [hunk 39]
```diff
@@ -0,0 +1,221 @@
...
+            hidden_params: RUBY_SECURE_PARAMS,
+            state: 'DISABLED'
+          )
+
+          associate_output_stage ads
+          add_process = project.add.process
+          add_process.create_schedule(
+            ruby_schedule,
+            '',
+            dataload_datasets: [ DATASET_IDENTIFIER ],
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Style/SpaceInsideBrackets: Space inside square brackets detected.

File: spec/helpers/appstore_project_helper.rb [hunk 40]
```diff
@@ -0,0 +1,221 @@
...
+                                    position_y: i['position']['y'],
+                                    size_x: i['size']['width'],
+                                    size_y: i['size']['height'])
+              end
+
+              tab.add_report_item(report: GoodData::Report['report.report.contains.ca'],
+                                  position_x: 0,
+                                  position_y: 0,
+                                  size_x: 640,
+                                  size_y:100)
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Style/SpaceAfterColon: Space missing after colon.

File: spec/helpers/appstore_project_helper.rb [hunk 41]
```diff
@@ -0,0 +1,221 @@
...
+        dataset.save
+
+        fact = @project.facts(FACT_IDENTIFIER)
+        fact.add_tag('fact')
+        fact.save
+      end
+
+      def ensure_user(login, domain)
+        user = domain.users(login)
+        unless user
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Style/IfUnlessModifier: Favor modifier `unless` usage when having a single-line body. Another good alternative is the usage of control flow `&&`/`||`.

File: spec/helpers/appstore_project_helper.rb [hunk 42]
```diff
@@ -0,0 +1,221 @@
...
+          user = domain.add_user(login: login)
+        end
+        @project.add_user(user, 'Viewer', domain: domain)
+        user
+      end
+    end
+  end
+end
+
```
Comment (yenkins) [inline spec/helpers/appstore_project_helper.rb]: Style/TrailingBlankLines: 1 trailing blank lines detected.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/667/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/468/ : FAILURE in 42s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/776/ : SUCCESS in 37s

**Comment by kubamahnert** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/668/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/469/ : FAILURE in 32s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/777/ : SUCCESS in 34s

**Comment by kubamahnert** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/669/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/470/ : FAILURE in 32s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/778/ : SUCCESS in 36s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/670/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/471/ : SUCCESS in 32s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/779/ : SUCCESS in 33s

**Comment by kubamahnert** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/671/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1175/ : FAILURE in 47m 27s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/933/ : UNSTABLE in 29m 02s

**Comment by kubamahnert** [comment]
extended test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/672/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/472/ : SUCCESS in 2m 15s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/780/ : SUCCESS in 2m 38s

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/673/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1176/ : SUCCESS in 53m 14s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/934/ : SUCCESS in 37m 09s

**Comment by panjan** [comment]
ok to merge

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-merge-compliance-check/259/ : SUCCESS in 47s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/473/ : SUCCESS in 30s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/781/ : SUCCESS in 35s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1177/ : SUCCESS in 51m 15s
- https://ci.intgdc.com/job/gooddata-ruby-lcm-integration-tests-zuul/935/ : SUCCESS in 30m 52s

**Pull request merged**